namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala

```diff
@@ -100,14 +100,16 @@ object ThriftNamerInterface {
     protected[this] def nextStamp(): Stamp
 
     protected[this] final def update(v: Try[T]): Unit = {
-      val obs = Observation(nextStamp(), v)
-      val promise = synchronized {
-        val previous = pending
-        current = Some(obs)
-        pending = new Promise[Observation]
-        previous
+      if (!current.exists(_.value == v)) {
+        val obs = Observation(nextStamp(), v)
+        val promise = synchronized {
+          val previous = pending
+          current = Some(obs)
```

Comment:
this is doing a read and then a write on current, which I believe needs to be synchronized.  Two threads could read `current`, and evaluate to false, and then both of them would create a new Observation, and an arbitrary one would win, causing setValue to be called twice, which throws an exception.